client/verta/verta/pipeline/_registered_pipeline.py

```diff
@@ -0,0 +1,174 @@
+        self._id = self._registered_model_version.id
+        self._pipeline_graph = pipeline_graph
+
+    def __repr__(self):
+        return "\n".join(
+            (
+                "RegisteredPipeline:",
+                f"pipeline name: {self.name}",
+                f"pipeline id: {self.id}",
+                f"\n{self._pipeline_graph}",
+            )
+        )
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def id(self):
+        return self._id
+
+    @property
+    def pipeline_graph(self):
+        return self._pipeline_graph
+
+    def copy_graph(self) -> PipelineGraph:
+        """
+        Return a shallow copy of the PipelineGraph of this pipeline. RegisteredPipeline
+        objects are immutable once registered with Verta. This function returns
+        a PipelineGraph object that can be modified and used to create and register
+        a new RegisteredPipeline.
+        """
+        return copy.copy(self.pipeline_graph)
+
+    def _log_pipeline_definition_artifact(self) -> None:
+        """
+        Log the pipeline definition as an artifact of the registered model version.
+        """
+        self._registered_model_version.log_artifact(
+            "pipeline.json", self._to_pipeline_definition()
+        )
+
+    def _get_pipeline_definition_artifact(self) -> Dict[str, Any]:
+        """
+        Get the pipeline definition artifact from the registered model version.
+        """
+        return self._registered_model_version.get_artifact("pipeline.json")
+
+    def _to_pipeline_definition(self) -> Dict[str, Any]:
+        """
+        Create a complete pipeline definition dict from a name and PipelineGraph.
+        Used in conjunction with the client function for creating a registered
+        pipeline from a pipeline graph.
+        """
+        return {
+            "pipeline_version_id": self.id,
+            "graph": self.pipeline_graph._to_graph_definition(),
+            "predecessors": self.pipeline_graph._to_steps_definition(),
+        }
+
+    def _to_pipeline_configuration(
+        self, pipeline_resources: Optional[Dict[str, Resources]] = None
+    ) -> Dict[str, Any]:
+        """
+        Build a pipeline configuration dict for this pipeline. The
+        `env` and `build` keys are not included in the configuration
+        resulting in default values being used by the backend.
+
+        Parameters
+        ----------
+        pipeline_resources : Resources
+
+        Returns
+        -------
+        Dictionary representation of a pipeline configuration.
+        """
+        steps = list()
+        for step in self.pipeline_graph.steps:
+            step_config = {
+                "name": step.name,
+            }
+            if pipeline_resources:
+                step_res = pipeline_resources.get(step.name, None)
+                if step_res:
+                    step_config["resources"] = pipeline_resources.pop(
```

Comment:
I'd _strongly_ not recommend mutating function arguments like this. It's an easy way to encounter bugs on the caller side. We should either `pipeline_resources = pipeline_resources.copy()`, or check the keys in a single loop at the beginning of the function.

Reply:
Also if we're really going with `pop()`, I think we could've done this two lines above

```python
step_res = pipeline_resources.pop(step.name, None)
```

Reply:
Refactored; no longer doing this.